### ISSUE TITLE
Add "How to distribute gradient computation?" document

### ISSUE THREAD

**Issue by rlouf** [blackjax-devs/blackjax#319]

**Comment by antotocar34** [comment]
I've been having a go at implementing this.

The problem I'm encountering is that the `VIAlgorithm` API is not really suitable for this algorithm.
SVGD doesn't have an `approximation` and `sampling` phase. In fact no random sampling happens
at all, SVGD is a deterministic algorithm.

SVGD is really just a gradient descent type procedure, which means the user will probably want the following
exposed to them:

1. A function that does one step of the optimization procedure
2. A way to add an arbitrary optimizer (adam, adagrad, etc...)
3. A way to take batch (in parallel if desired) gradients of the log posterior

Here are my thoughts:
1. 
The existing mcmc kernel API is almost suitable for SVGD, except that there is no need for RNG keys 
since the algorithm is deterministic.
2.
The `optax` library provides many state of the art optimizers, so it would be nice to have some compatibility with it.
We could define an Optimizer Protocol that implements the same methods as optax optimizers
(namely `init` and `update`), and have those users who want a custom optimizer adhere to this Protocol.
3.
I believe this can be done as suggested in #319.

### Proposal Implementation
(To be clear, this would not be the final implementation, just to pin down ideas)

In `./blackjax/base.py`
```python
class SvgdStepFn(Protocol):
    def __call__(self, state: State) -> State:
        ...

class SVGDAlgorithm(NamedTuple):
    init: InitFn
    step: SvgdStepFn
```

In `./blackjax/svgd.py`
```python
class SVGDState(NamedTuple):
    position: PyTree
    kernel_parameters: PyTree
    opt_state: Any

def init(initial_position: PyTree, kernel_parameters: PyTree) -> SVGDState:
    opt_state = optimizer.init(initial_position)
    return SVGDState(initial_position, kernel_parameters, opt_state)

def step(state: SVGDState, log_p: Callable, kernel: Callable, optimizer: Optimizer) -> SVGDState:
    # Computation where we call kernel as kernel(x,y, **state.kernel_params)
    ...
```

In `./blackjax/kernels.py`
```python
class Optimizer(Protocol):
    init: Callable
    update: Callable 

class svgd:
    step = staticmethod(vi.svgd.step)
    def __new__( cls, log_p: Callable, kernel: Callable, optimizer: Optimizer) -> SVGDAlgorithm:

        def init_fn(initial_position: PyTree, kernel_parameters: Dict):
            return cls.init(initial_position, state.kernel_parameters)

        def step_fn(
                state: SVGDState,
                ):
            return cls.step(state, log_p, kernel, optimizer)

    return SVGDAlgorithm(init_fn, step_fn)
```

#### An example from the user's point of view
```python
svgd = svgd(log_p, kernel, optimizer) # example optimizer: optax.adam(0.1)
initial_state = svgd.init(initial_position)
step = jax.jit(svgd.step)(state)

def update_kernel_parameters(state: SVGDState) -> SVGDState:
    position, kernel_parameters, opt_state = state
    # f could be the median heuristic for example
    return SVGDState(position, f(kernel_parameters, position), opt_state)


num_iterations = 5000
for _ in range(num_interations):
    state = svgd.step(state)
    state = update_kernel_parameters(state)
```

Let me know your thoughts :)
